### PR TITLE
Add routed (MoE-style) N-gram embedding with sub-table load balancing

### PR DESCRIPTION
**Pull request by Ngk03**
### Description

Adds LongCat-style N-gram embedding scaling in two forms, a dense one and a routed (MoE-style) one, plus two ways of balancing load across the routed sub-tables. Off by default.

Both forms read, at each position, an embedding keyed on the N-gram ending at that position (for `N=3`, the bigram and the trigram), project it to `hidden_size`, and average it into the word embedding. The transformer stack above is untouched.

New modules under `src/paddlefleet/models/common/embeddings/`:

- `ngram_embedding.py` — the dense form. The row is a pure hash of the N-gram, over `ngram_emb_split_num` independent sub-tables per order, each with its own table and projection.
- `ngram_moe_embedding.py` — the routed form. A router picks which sub-tables of each order to read and the hash picks the row inside them, so the coarse part of the address is learned while the fine part stays hashed. The router sees only tokens up to the current position. The selected reads are combined before the projection, so the number of projections does not grow with the number of sub-tables. An optional gram-level gate can reweight bigram against trigram.
- `ngram_bias_callback.py` — the per-step update of the aux-loss-free balancing bias.

`ngram_moe_balance_type` selects how sub-table load is balanced: `"noaux_bias"` (default; the aux-loss-free bias of DeepSeek-V3, which requires the callback above to be registered on the trainer), `"aux_loss"` (the classic gradient-based auxiliary loss), or `"none"`. Both follow this repo's existing MoE conventions: the bias enters the top-k selection while the combine weights come from the unbiased router, as in `MoERouter._topk_noaux_tc`, and the auxiliary loss has the same form and normalisation as `MoERouter._cal_aux_loss`, applied per N-gram order. Config fields are documented on `TransformerConfig`.

**Parallelism.** Written for TP = 1: the tables, their projections and the router are not tensor-parallel sharded, so under TP > 1 every TP rank would hold a replica and recompute the same signal. Data parallel and sharding are supported; the usage counts behind the aux-loss-free bias are all-reduced over the groups whose ranks hold different tokens, following `moe/qb_callback.py`.

**Non-invasiveness.** `LanguageModelEmbedding` builds either variant only when `ngram_embedding_enabled` is set, and fuses the signal in `[b s h]` layout before any transpose or sequence-parallel scatter. With the feature off, every code path, every parameter and the resulting state dict are exactly what they were, so the architecture is switchable purely from configuration and existing checkpoints keep loading.

Reference: Meituan LongCat-Flash-Lite (https://huggingface.co/meituan-longcat/LongCat-Flash-Lite)

### 是否引起精度变化

否。`ngram_embedding_enabled` 默认为 `False`。

### CODE REVIEW

File: src/paddlefleet/models/common/embeddings/ngram_moe_embedding.py [hunk 1]
```diff
@@ -0,0 +1,656 @@
...
+
+    def _compute_shifted_ids(self, input_ids: Tensor) -> dict:
+        """Right-shifted copies of input_ids, one per offset (same as baseline)."""
+        shifted_ids = {}
+        B, S = input_ids.shape
+        for k in range(1, self.n):
+            shifted = paddle.full(
+                [B, k], self.pad_token_id, dtype=input_ids.dtype
+            )
+            shifted = paddle.concat([shifted, input_ids[:, : S - k]], axis=1)
```
Comment (risemeup1111) [inline src/paddlefleet/models/common/embeddings/ngram_moe_embedding.py:384]: P0 `_compute_shifted_ids` 在 `S < k` 时会产生超过原序列长度的张量：例如 `S=1,k=2` 时 `input_ids[:, :S-k]` 仍与 `[B,k]` padding 拼成 `[B,2]`。随后 embedding 输出被广播到错误的序列长度，`LanguageModelEmbedding.forward` 还会继续把它与 `[B,1,H]` 相加，最终返回 shape 错误（dense 版本的同名实现也有相同问题）。请把右移结果显式截断/补齐到 `S`，并覆盖 `S=1`、`S<N`。

File: src/paddlefleet/models/common/embeddings/ngram_moe_embedding.py [hunk 2]
```diff
@@ -0,0 +1,656 @@
...
+                num_tables=self.tables_per_order,
+                router_dim=_router_dim,
+                width=_router_width,
+            )
+
+        # Sub-table load balancing.  Exactly one style is active.
+        self.balance_type = str(
+            getattr(config, "ngram_moe_balance_type", "noaux_bias")
+        ).lower()
+        assert self.balance_type in ("aux_loss", "noaux_bias", "none"), (
```
Comment (risemeup1111) [inline src/paddlefleet/models/common/embeddings/ngram_moe_embedding.py:322]: P1 `active_tables`、`table_rows`、`emb_dim` 和 `balance_type` 的非法配置都依赖 Python `assert`。在 `python -O` 下这些检查会被完全移除，未知 balance type 会静默走到无负载均衡路径，零/负尺寸也会延迟到 Paddle 算子才失败。请将这些约束放到 `TransformerConfig.__post_init__` 或构造入口，用 `ValueError` 显式拒绝，并为非法组合补测试。

File: src/paddlefleet/models/common/embeddings/ngram_moe_embedding.py [hunk 3]
```diff
@@ -0,0 +1,656 @@
...
+
+        # One flat table over all (order, sub-table) pairs: the routed address is
+        # then already the embedding row index, with no per-sub-table gather.
+        # Materialized as ``num_shards`` slices along the feature dim so that no
+        # single parameter exceeds int32 numel (see _MAX_PARAM_NUMEL).
+        self.num_shards = _num_table_shards(total, self.emb_dim)
+        self.shard_dim = self.emb_dim // self.num_shards
+        # fan_in / fan_out pinned to the unsharded shape, so the initial
+        # distribution does not depend on how many slices are used.
+        self.table_shards = nn.LayerList(
```
Comment (risemeup1111) [inline src/paddlefleet/models/common/embeddings/ngram_moe_embedding.py:271]: P1 路由表和投影在这里用普通 `nn.Embedding`/`nn.Linear` 构造；当 `tensor_model_parallel_size > 1` 时，每个 TP rank 都会分配完整的 routed table、投影和 router，代码没有 TP 分片、collective、`is_distributed` 标记或 `sharded_state_dict`。因此启用 N-gram 的大表配置会在每个 rank 复制全部参数，内存按 TP 度数额外放大，并不符合本仓库的 TP 参数契约。请改用 TP-aware 分片/保存逻辑，或在配置层明确拒绝未支持的 TP 组合并加多卡覆盖。

File: src/paddlefleet/models/common/embeddings/ngram_moe_embedding.py [hunk 4]
```diff
@@ -0,0 +1,656 @@
...
+            [
+                nn.Embedding(
```
Comment (risemeup1111) [inline src/paddlefleet/models/common/embeddings/ngram_moe_embedding.py:273]: P1 新增的 table、projection 和 router 层没有使用 `config.params_dtype`。仓库的 bf16 配置会把 `VocabParallelEmbedding` 初始化为 bf16，但 Paddle 默认 dtype 仍是 fp32，因此这里的 `ngram_signal`/中间激活保持 fp32，随后在 `LanguageModelEmbedding.forward` 中与 bf16 word embedding 相加会改变输出 dtype并增加显存，O2 也无法按注释中的预期工作。请按 `params_dtype` 创建/转换这些层（仅 balance bias 保持 fp32），并补 bf16/O2 覆盖。

File: src/paddlefleet/transformer/transformer_config.py [hunk 5]
```diff
@@ -907,6 +907,151 @@ class TransformerConfig(ModelParallelConfig):
     fp8_weight_quant_format: str = "32x32"
     """Quantization format for quantizing weights, options are 32x32 and 1x32. Currently only used in SonicMoE."""
 
+    ####################
+    # N-gram Embedding (LongCat-style)
+    ####################
+    ngram_embedding_enabled: bool = False
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_config.py:913]: P1 当前 head 新增了两套 embedding、路由/平衡回调和一组配置字段，但 PR diff 没有任何 `tests/` 变更。短序列 shape、非法配置、bf16、TP/SP 参数布局以及 callback 的跨组归约都没有回归保护，以上问题无法由 CI 捕获。请至少增加单卡 forward/backward 与异常路径测试，并增加 TP/SP/回调通信的多卡覆盖。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 规则来源 | 状态 |
| --- | --- | --- | --- | --- |
| 1 | [短序列右移](https://github.com/PaddlePaddle/PaddleFleet/pull/1838#discussion_r3838287329) | P0 | [仓库规则：基础评审规则](https://github.com/PaddlePaddle/PaddleFleet/blob/a40748de84cd4a3611f92d3fe03a542f12653a15/.agents/skills/ai-review/references/base-rules.md) | 🚧 |
| 2 | [配置校验](https://github.com/PaddlePaddle/PaddleFleet/pull/1838#discussion_r3838289751) | P1 | [仓库规则：基础评审规则](https://github.com/PaddlePaddle/PaddleFleet/blob/a40748de84cd4a3611f92d3fe03a542f12653a15/.agents/skills/ai-review/references/base-rules.md)<br>[仓库规则：用户配置开关规则](https://github.com/PaddlePaddle/PaddleFleet/blob/a40748de84cd4a3611f92d3fe03a542f12653a15/.agents/skills/ai-review/references/user-configuration-rules.md) | 🚧 |
| 3 | [TP 参数布局](https://github.com/PaddlePaddle/PaddleFleet/pull/1838#discussion_r3838290524) | P1 | [仓库规则：基础评审规则](https://github.com/PaddlePaddle/PaddleFleet/blob/a40748de84cd4a3611f92d3fe03a542f12653a15/.agents/skills/ai-review/references/base-rules.md) | 🚧 |
| 4 | [参数 dtype](https://github.com/PaddlePaddle/PaddleFleet/pull/1838#discussion_r3838291357) | P1 | [仓库规则：基础评审规则](https://github.com/PaddlePaddle/PaddleFleet/blob/a40748de84cd4a3611f92d3fe03a542f12653a15/.agents/skills/ai-review/references/base-rules.md) | 🚧 |
| 5 | [回归测试覆盖](https://github.com/PaddlePaddle/PaddleFleet/pull/1838#discussion_r3838292114) | P1 | [仓库规则：基础评审规则](https://github.com/PaddlePaddle/PaddleFleet/blob/a40748de84cd4a3611f92d3fe03a542f12653a15/.agents/skills/ai-review/references/base-rules.md)<br>[仓库规则：用户配置开关规则](https://github.com/PaddlePaddle/PaddleFleet/blob/a40748de84cd4a3611f92d3fe03a542f12653a15/.agents/skills/ai-review/references/user-configuration-rules.md) | 🚧 |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.6-sol 默认推理级别, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
新 head 仅调整注释，原有五项 P0/P1 问题均未解决，维持修改请求。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.6-sol 默认推理级别, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
新 head 仅调整文档和注释，原有五项 P0/P1 问题均未解决，维持修改请求。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.6-sol 默认推理级别, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
新 head 仅调整注释和配置说明，未修复原有五项 P0/P1 问题，维持修改请求。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.6-sol 默认推理级别, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/models/common/embeddings/ngram_moe_embedding.py (hunk 4 above)
Comment (risemeup1111) [reply src/paddlefleet/models/common/embeddings/ngram_moe_embedding.py:273]: 新提交在融合前将 `ngram_signal` 转为 word embedding 的 dtype，解决了最终相加的 dtype mismatch；但 `NgramEmbedding` / `NgramMoeEmbedding` 内的 Embedding、Linear 和 router 仍按默认 dtype 构造，参数与中间激活仍未遵循 `config.params_dtype`，因此 bf16/O2 的显存与参数 dtype 问题仍未完整修复。

File: src/paddlefleet/models/common/embeddings/ngram_moe_embedding.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/models/common/embeddings/ngram_moe_embedding.py:322]: 新提交新增了 `ngram_emb_neighbor_num` / `ngram_emb_split_num` 与 routed N-gram 的 `assert`，但这些校验在 `python -O` 下仍会被移除；`active_tables`、行数、维度和 `balance_type` 也仍未改为 `ValueError` 配置校验，因此该 P1 继续未解决。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
新 head 部分调整了 dtype 融合和配置断言，但原有短序列 shape、显式配置校验、TP 参数布局、完整 params_dtype 支持及回归测试五项 P0/P1 仍未解决，维持修改请求。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.6-sol 默认推理级别, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
新 head 仅调整 callback 注释，原有五项 P0/P1 问题均未解决，维持修改请求。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.6-sol 默认推理级别, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
新 head 仅调整 callback 文档注释，原有五项 P0/P1 问题均未解决，维持修改请求。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.6-sol 默认推理级别, please check the suggestions carefully.</sup>
</div>
